applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -39,830 +37,673 @@
-        ModelPart& r_model_part,
-        TSystemMatrixType& A,
-        TSystemVectorType& b) override
+        ModelPart &rModelPart,
+        TSystemMatrixType &rA,
+        TSystemVectorType &rb,
+        const bool BuildLHS = true,
+        const bool BuildRHS = true)
     {
         KRATOS_TRY
-        if(!pScheme)
-            KRATOS_ERROR << "No scheme provided!";
-
-        //getting the elements from the model
-        ElementsArrayType& pElements = r_model_part.Elements();
 
-        //getting the array of the conditions
-        ConditionsArrayType& ConditionsArray = r_model_part.Conditions();
+        KRATOS_ERROR_IF(!pScheme) << "No scheme provided!" << std::endl;
 
-        //resetting to zero the vector of reactions
-        TSparseSpace::SetToZero(*BaseType::mpReactionsVector);
+        const int nelements = static_cast<int>(rModelPart.Elements().size());
+        const int nconditions = static_cast<int>(rModelPart.Conditions().size());
+        // Resetting to zero the vector of reactions
+        TSparseSpace::SetToZero(*BaseType::mpReactionsVector); // TODO: Check if required
 
-        //contributions to the system
-        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0,0);
+        // Contributions to the system
+        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0, 0);
         LocalSystemVectorType RHS_Contribution = LocalSystemVectorType(0);
 
-        //vector containing the localization in the system of the different
-        //terms
-        Element::EquationIdVectorType EquationId;
-
-        //			int rank = A.Comm().MyPID(); //getting the processor Id
+        //vector containing the localization in the system of the different terms
+        Element::EquationIdVectorType equation_ids_vector;
+        ProcessInfo &r_current_process_info = rModelPart.GetProcessInfo();
+        ModelPart::ElementsContainerType::iterator el_begin = rModelPart.ElementsBegin();
+        ModelPart::ConditionsContainerType::iterator cond_begin = rModelPart.ConditionsBegin();
 
-        ProcessInfo& CurrentProcessInfo = r_model_part.GetProcessInfo();
         // assemble all elements
-        for (typename ElementsArrayType::ptr_iterator it=pElements.ptr_begin(); it!=pElements.ptr_end(); ++it)
+        for (int k = 0; k < nelements; k++)
         {
-            //calculate elemental contribution
-            pScheme->CalculateSystemContributions(*it,LHS_Contribution,RHS_Contribution,EquationId,CurrentProcessInfo);
+            ModelPart::ElementsContainerType::iterator it = el_begin + k;
 
-            //assemble the elemental contribution
-            TSparseSpace::AssembleLHS(A,LHS_Contribution,EquationId);
-            TSparseSpace::AssembleRHS(b,RHS_Contribution,EquationId);
+            //detect if the element is active or not. If the user did not make any choice the element
+            //is active by default
+            bool element_is_active = true;
```

Comment:
It should only call the Is() if the first is true.
```suggestion
            const bool element_is_active = it->IsDefined(ACTIVE) && it->Is(ACTIVE);
```

Reply:
I second the suggestion of @msandre here, but the check should actually be
```C++
const bool element_is_active = !(it->IsDefined(ACTIVE)) || it->Is(ACTIVE);
```